packages/gatsby/src/redux/persist.ts

```diff
@@ -1,12 +1,166 @@
+import path from "path"
 import v8 from "v8"
-import { readFileSync, writeFileSync } from "fs-extra"
-import { ICachedReduxState } from "./types"
+import {
+  existsSync,
+  mkdtempSync,
+  readFileSync,
+  removeSync,
+  renameSync,
+  writeFileSync,
+} from "fs-extra"
+import { IReduxNode, ICachedReduxState } from "./types"
+import { sync as globSync } from "glob"
 
-const file = (): string => `${process.cwd()}/.cache/redux.state`
+const getLegacyCacheFile = (): string =>
+  // TODO: remove this legacy stuff in v3 (fairly benign change but still)
+  // This is a function for the case that somebody does a process.chdir (#19800)
+  path.join(process.cwd(), `.cache/redux.state`)
 
-export const readFromCache = (): ICachedReduxState =>
-  v8.deserialize(readFileSync(file()))
+const getReduxCacheFolder = (): string =>
+  // This is a function for the case that somebody does a process.chdir (#19800)
+  path.join(process.cwd(), `.cache/redux`)
 
-export const writeToCache = (contents: ICachedReduxState): void => {
-  writeFileSync(file(), v8.serialize(contents))
+function reduxRestFile(dir: string): string {
+  return path.join(dir, `redux.rest.state`)
+}
+function reduxChunkFilePrefix(dir: string): string {
+  return path.join(dir, `redux.node.state_`)
+}
+
+function readFromLegacyCache(): ICachedReduxState {
+  return v8.deserialize(readFileSync(getLegacyCacheFile()))
+}
+
+export function readFromCache(): ICachedReduxState {
+  // The cache is stored in two steps; the nodes in chunks and the rest
+  // First we revive the rest, then we inject the nodes into that obj (if any)
+  // Each chunk is stored in its own file, this circumvents max buffer lengths
+  // for sites with a _lot_ of content. Since all nodes go into a Map, the order
+  // of reading them is not relevant.
+
+  const reduxCacheFolder = getReduxCacheFolder()
+
+  if (!existsSync(reduxCacheFolder)) {
+    return readFromLegacyCache()
+  }
+
+  const obj: ICachedReduxState = v8.deserialize(
+    readFileSync(reduxRestFile(reduxCacheFolder))
+  )
+
+  // Note: at 1M pages, this will be 1M/chunkSize chunks (ie. 1m/10k=100)
+  const chunks = globSync(
+    reduxChunkFilePrefix(reduxCacheFolder) + `*`
+  ).map(file => v8.deserialize(readFileSync(file)))
+
+  const nodes: [string, IReduxNode][] = [].concat(...chunks)
+
+  if (chunks.length) {
+    obj.nodes = new Map(nodes)
+  }
```

Comment:
Core creates some nodes on its own (things like `Site` node type etc - https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/internal-plugins/internal-data-bridge/gatsby-node.js#L98-L107), so there always should be at least one node cached (if we have any cache that is). 

Would it be good idea to have `else` path here in case something went wrong and there are no nodes read from `.cache/redux` (in this code path `.cache/redux` dir exists) and throw? This should make it more proofed and add some very crude "state consistency check"

_This might be wrong place to handle this concern (as it relies about knowledge that core does indeed always create at least one node)_